nifi-nar-bundles/nifi-hive-bundle/nifi-hive-processors/src/main/java/org/apache/nifi/processors/hive/PutHiveStreaming.java

```diff
@@ -0,0 +1,657 @@
+                .withAutoCreatePartitions(autoCreatePartitions)
+                .withMaxOpenConnections(maxConnections)
+                .withHeartBeatInterval(heartbeatInterval);
+
+        if (SecurityUtil.isSecurityEnabled(hiveConfig)) {
+            final String principal = context.getProperty(kerberosProperties.getKerberosPrincipal()).getValue();
+            final String keyTab = context.getProperty(kerberosProperties.getKerberosKeytab()).getValue();
+
+            log.info("Hive Security Enabled, logging in as principal {} with keytab {}", new Object[]{principal, keyTab});
+            try {
+                ugi = hiveConfigurator.authenticate(hiveConfig, principal, keyTab, TICKET_RENEWAL_PERIOD, log);
+            } catch (AuthenticationFailedException ae) {
+                throw new ProcessException("Kerberos authentication failed for Hive Streaming", ae);
+            }
+            log.info("Successfully logged in as principal {} with keytab {}", new Object[]{principal, keyTab});
+            options = options.withKerberosPrincipal(principal).withKerberosKeytab(keyTab);
+        }
+
+        allWriters = new ConcurrentHashMap<>();
+        String timeoutName = "put-hive-streaming-%d";
+        this.callTimeoutPool = Executors.newFixedThreadPool(1,
+                new ThreadFactoryBuilder().setNameFormat(timeoutName).build());
+
+        sendHeartBeat.set(true);
+        heartBeatTimer = new Timer();
+        setupHeartBeatTimer();
+    }
+
+    @Override
+    public void onTrigger(ProcessContext context, ProcessSession session) throws ProcessException {
+        FlowFile flowFile = session.get();
+        if (flowFile == null) {
+            return;
+        }
+
+        final ComponentLog log = getLogger();
+        try {
+            final List<String> partitionColumnList;
+            String partitionColumns = context.getProperty(PARTITION_COLUMNS).getValue();
+            if (StringUtils.isEmpty(partitionColumns)) {
+                partitionColumnList = Collections.emptyList();
+            } else {
+                String[] partitionCols = partitionColumns.split(",");
+                partitionColumnList = new ArrayList<>(partitionCols.length);
+                for (String col : partitionCols) {
+                    partitionColumnList.add(col.trim());
+                }
+            }
+
+            // Store the original class loader, then explicitly set it to this class's classloader (for use by the Hive Metastore)
+            ClassLoader originalClassloader = Thread.currentThread().getContextClassLoader();
+            Thread.currentThread().setContextClassLoader(this.getClass().getClassLoader());
+
+            int recordCount = 0;
+            final List<HiveStreamingRecord> records = new LinkedList<>();
+
+            session.read(flowFile, in -> {
+
+                try (final DataFileStream<GenericRecord> reader = new DataFileStream<>(in, new GenericDatumReader<GenericRecord>())) {
```

Comment:
I know using Java 8 lambdas can be addictive, . . . and therefore dangerous as it may make code less readable. And IMHO this is the case. I would personally put the code in try/catch ins a separate method and then
```
 session.read(flowFile, in -> fooMethod(in))
```